src/v/cloud_storage/topic_manifest.cc

```diff
@@ -12,21 +12,193 @@
 
 #include <boost/lexical_cast.hpp>
+#include <boost/lexical_cast/try_lexical_convert.hpp>
 #include <fmt/ostream.h>
+#include <rapidjson/encodings.h>
+#include <rapidjson/error/en.h>
 #include <rapidjson/istreamwrapper.h>
 #include <rapidjson/ostreamwrapper.h>
+#include <rapidjson/reader.h>
 
 #include <chrono>
+#include <optional>
+#include <stdexcept>
+#include <string_view>
 
 namespace cloud_storage {
+
+struct topic_manifest_handler
+  : public rapidjson::
+      BaseReaderHandler<rapidjson::UTF8<>, topic_manifest_handler> {
+    using key_string = ss::basic_sstring<char, uint32_t, 31>;
+    bool StartObject() {
+        switch (_state) {
+        case expectManifestStart:
+            _state = expectKey;
+            return true;
+        case expectKey:
+        case expectValue:
+            return false;
+        }
+    }
+
+    bool Key(const char* str, rapidjson::SizeType length, bool /*copy*/) {
+        switch (_state) {
+        case expectKey:
+            _key = key_string(str, length);
+            _state = expectValue;
+            return true;
+        case expectManifestStart:
+        case expectValue:
+            return false;
+        }
+    }
+
+    bool String(const char* str, rapidjson::SizeType length, bool /*copy*/) {
+        std::string_view sv(str, length);
+        switch (_state) {
+        case expectValue:
+            if ("namespace" == _key) {
+                _namespace = model::ns(ss::sstring(sv));
+            } else if ("topic" == _key) {
+                _topic = model::topic(ss::sstring(sv));
+            } else if ("compression" == _key) {
+                model::compression compression_res;
+                if (boost::conversion::try_lexical_convert(
+                      sv, compression_res)) {
+                    _properties.compression = compression_res;
+                } else {
+                    return false;
+                }
+            } else if ("cleanup_policy_bitflags" == _key) {
+                model::cleanup_policy_bitflags cleanup_policy_bitflags_res;
+                if (boost::conversion::try_lexical_convert(
+                      sv, cleanup_policy_bitflags_res)) {
+                    _properties.cleanup_policy_bitflags
+                      = cleanup_policy_bitflags_res;
+                } else {
+                    return false;
+                }
+            } else if ("compaction_strategy" == _key) {
+                model::compaction_strategy compaction_strategy_res;
+                if (boost::conversion::try_lexical_convert(
+                      sv, compaction_strategy_res)) {
+                    _properties.compaction_strategy = compaction_strategy_res;
+                }
+            } else if ("timestamp_type" == _key) {
+                model::timestamp_type timestamp_type_res;
+                if (boost::conversion::try_lexical_convert(
+                      sv, timestamp_type_res)) {
+                    _properties.timestamp_type = timestamp_type_res;
```

Comment:
do you have tests that cover the cases when timestamp_type (and others like compaction_strategy etc...) have invalid values?

Reply:
I have a `wrong_enum_throws` test for invalid `compression` value. The test just expects `std::runtime_error` which is also thrown for missing required fields, so maybe I should test for the json with all values present and just some value being invalid.

Reply:
Will fix tests